compiler/src/iree/compiler/Codegen/Dialect/GPU/Transforms/ConcretizeMmaShapes.cpp

```diff
@@ -0,0 +1,133 @@
+                                           lhsReassociations)
+            .getResult();
+    Value concreteRhs =
+        rewriter
+            .create<tensor::ExpandShapeOp>(loc, rhsConcreteType, mmaOp.getRhs(),
+                                           rhsReassociations)
+            .getResult();
+
+    std::optional<DenseI64ArrayAttr> lhsPerm, rhsPerm, accPerm;
+    if (mmaOp.getLhsPermutation().has_value())
+      lhsPerm = mmaOp.getLhsPermutationAttr();
+    if (mmaOp.getRhsPermutation().has_value())
+      rhsPerm = mmaOp.getRhsPermutationAttr();
+    if (mmaOp.getAccPermutation().has_value())
+      accPerm = mmaOp.getAccPermutationAttr();
+
+    rewriter.replaceOpWithNewOp<IREE::GPU::MultiMmaOp>(
+        mmaOp, concreteLhs, concreteRhs, mmaOp.getAcc(),
+        mmaOp.getIndexingMaps(), mmaOp.getIteratorTypes(), mmaOp.getKind(),
+        lhsPerm, rhsPerm, accPerm);
+
+    return success();
+  }
+};
+
+struct ConcretizeMmaResultShape final
+    : OpRewritePattern<IREE::GPU::MultiMmaOp> {
+  using OpRewritePattern::OpRewritePattern;
+  LogicalResult matchAndRewrite(IREE::GPU::MultiMmaOp mmaOp,
+                                PatternRewriter &rewriter) const override {
+    IREE::GPU::MmaInterfaceAttr kind = mmaOp.getKind();
+    SmallVector<ReassociationIndices> accReassociations;
+    RankedTensorType accConcreteType;
+    if (failed(kind.materializeOperandConcreteShape(
+            rewriter, IREE::GPU::MMAFragment::Acc, mmaOp.getAcc(),
+            mmaOp.getAccPermutation(), accReassociations, accConcreteType))) {
+      return failure();
+    }
+    Location loc = mmaOp->getLoc();
+    Value concreteAcc =
+        rewriter
+            .create<tensor::ExpandShapeOp>(loc, accConcreteType, mmaOp.getAcc(),
+                                           accReassociations)
+            .getResult();
+
+    std::optional<DenseI64ArrayAttr> lhsPerm, rhsPerm, accPerm;
+    if (mmaOp.getLhsPermutation().has_value())
+      lhsPerm = mmaOp.getLhsPermutationAttr();
+    if (mmaOp.getRhsPermutation().has_value())
+      rhsPerm = mmaOp.getRhsPermutationAttr();
+    if (mmaOp.getAccPermutation().has_value())
+      accPerm = mmaOp.getAccPermutationAttr();
+
+    auto concreteMmaOp = rewriter.create<IREE::GPU::MultiMmaOp>(
+        loc, mmaOp.getLhs(), mmaOp.getRhs(), concreteAcc,
+        mmaOp.getIndexingMaps(), mmaOp.getIteratorTypes(), mmaOp.getKind(),
+        lhsPerm, rhsPerm, accPerm);
+
+    rewriter.replaceOpWithNewOp<tensor::CollapseShapeOp>(
+        mmaOp, mmaOp.getAccType(), concreteMmaOp.getResult(),
+        accReassociations);
+
+    return success();
+  }
+};
+
+void ConcretizeMmaShapesPass::runOnOperation() {
+  MLIRContext *context = &getContext();
+  auto funcOp = getOperation();
+
+  RewritePatternSet patterns(context);
+  if (concretizeInputs) {
+    patterns.insert<ConcretizeMmaInputShapes>(context);
+  }
+  if (concretizeResult) {
+    patterns.insert<ConcretizeMmaResultShape>(context);
+  }
```

Comment:
optional: It might look cleaner to combine the patterns by passing in a `IREE::GPU::MMAFragment` as a pattern argument and pick which value to expand based on the fragment.